quinn/src/endpoint.rs

```diff
@@ -107,7 +107,8 @@ impl Endpoint {
         Self::new_with_runtime(config, server_config, Box::new(socket), Arc::new(runtime))
     }
 
-    fn new_with_runtime(
+    /// Construct an endpoint with arbitrary configuration and pre-constructed async socket.
+    pub fn new_with_runtime(
```

Comment:
This was needed to get my copy of the quinn examples working with boringssl.

Reply:
Can you elaborate on why?

Reply:
I'm trying to make helper methods, similar to what quinn has for rustls. For example, this is a boring version of a helper method to create a client endpoint:

```rust
pub fn client_endpoint(addr: SocketAddr) -> io::Result<quinn::Endpoint> {
        let socket = std::net::UdpSocket::bind(addr)?;
        let runtime = default_runtime()
            .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "no async runtime found"))?;
        quinn::Endpoint::new_with_runtime(
            default_endpoint_config(),
            None,
            runtime.wrap_udp_socket(socket)?,
            runtime,
        )
    }
```

For this to work, `Endpoint::new_with_runtime` has to be public. The other method `Endpoint::new_with_abstract_socket` requires that the caller has a concrete, unboxed `tokio` udp socket. I'm still pretty new to rust and I suspect I can make something work, but it seemed (at least initially) logical for `new_with_runtime` to be public.

Reply:
If you build the `std::net::UdpSocket`, it looks like you can just call `new()` instead of `new_with_runtime()`? It wraps the socket for you.

Reply:
Ok, I guess the last issue is that `default_runtime` returns an `Arc<dyn Runtime>`, but `new` requires an `impl Runtime`.  I can dereference the `Arc`, but that gives me a `&dyn Runtime`. Anyway, this seems wrong, because `new` wraps the value in another `Arc`, which would effictively move ownership IIUC.

I've seen other callsites of`new` and just directly pass in `TokioRuntime`. However, if we theoretically want to support other runtimes, using `default_runtime` seems better.

Still learning rust, so I'm probably missing something.

Reply:
Ah, I think this is a minor flaw in our API design. We could hack around the problem by defining `impl Runtime for Arc<dyn Runtime>`, but this causes gratuitous double indirection since it'll get re-`Arc`ed. I think we should replace `impl Runtime` with `Arc<dyn Runtime>` in the existing public APIs.

I think it's fine for test code to use `new` and hardcode `TokioRuntime` for now, and we can keep `new_with_rutnime` private and pursue API cleanup in future work.

Reply:
SGTM. done.